### ISSUE TITLE
Eta-reduce type aliases

### ISSUE THREAD

**Issue by rightfold** [purescript/purescript#2691]
That is, make:

    type T a b c = U a Int b c

syntactic sugar for:

    type T a = U a Int

This would partially applying such type aliases.

**Comment by garyb** [comment]
I like the idea of this. I don't think there's any reason to not do it?

**Comment by paf31** [comment]
I would rather warn with a suggestion than change the code on the users behalf.

**Comment by garyb** [comment]
Well, that was my initial reaction too, but if there's never ever a reason to explicitly eta-expand a synonym, why not eta-reduce without warning? They do mean the same thing in any sense that matters, so it doesn't really seem like we're changing anything. We're just being lenient in the form that the definition allows.

I guess one argument for making it warn instead, is that reducing automatically might make it less clear why some synonmys can be partially applied and others cannot. Like if you define `type T a = U a Int` then it seems more obvious why you can use it as `T a` and not `T`, but perhaps not so if it was defined as `type T a b c = U a Int b c`.

I'd be fine with it either way though really. I've been bitten by overly-expanded synonyms in other libraries (and made the mistake myself of course), so it would be great to catch those cases in the compiler.

**Comment by paf31** [comment]
Yes, I think a warning without changing the meaning makes it easier to explain the behavior.

**Comment by rightfold** [comment]
I suppose part of the documentation issue could be solved if you could have kind signatures e.g.

type T :: Type -> Type
type T = U

Sent from my iPhone.
